Topics/Tech_Stacks/Playwright.md

```diff
@@ -0,0 +1,169 @@
+# Playwright: Web Automation Made Easy
+## Table of Contents
+
+1. [What is Playwright?](#what-is-playwright)
+2. [Setup and Installation](#setup-and-installation)
+3. [Scripting](#scripting)
+4. [Headless Mode](#headless-mode)
+5. [Automatic Waiting](#automatic-waiting)
+6. [Browser Context](#browser-context)
+7. [Advantages Over Other Solutions](#advantages-over-other-solutions)
+8. [Limitations](#limitations)
+9. [Citation & Useful Links](#useful-links)
+
+## What is Playwright
+Playwright is an open-source end-to-end testing framework developed by Microsoft, designed specifically for testing modern web applications.
+
+Playwright enables automation of user interactions, including clicking on elements, filling in text fields, and more. It behaves like a user, waiting for elements to become actionable before performing actions and retrying assertions. Moreover, it features default parallel test execution which enhances speed therefore making the testing process exceptionally fast while aslo offering great browser support.
+
+## Setup and Installation
+To install Playwright run the following command:
+
+```
+npm init playwright@latest
+```
+It will download the browsers needed along with 
+```
+playwright.config.ts
+package.json
+package-lock.json
+tests/
+  example.spec.ts
+tests-examples/
+  demo-todo-app.spec.ts
+```
+To run the example tests do 
+```
+npx playwright test
+```
+
+Alternatively, you could install the Playwright extension on VSCode from the marketplace.
+<img width="714" alt="Screenshot 2024-03-17 at 3 42 39 PM" src="https://github.com/learning-software-engineering/learning-software-engineering.github.io/assets/113081946/e79ee478-9d94-49af-ab1a-fb53c6e06e5b">
+
+Once installed, open the command panel and type:
+```
+Install Playwright
+```
+Select Test: Install Playwright and then choose the browsers you would like to run your tests on. 
+<img width="714" alt="Screenshot 2024-03-17 at 3 44 23 PM" src="https://github.com/learning-software-engineering/learning-software-engineering.github.io/assets/113081946/7c2e3fa9-4224-4fc8-b24a-a2e9d20eb8f5">
+
+## Scripting 
+Playwright provides an API for scripting interactions with web pages. Here's an example of how to open a browser, navigate to a website, and perform actions:
+```
+const { chromium } = require('playwright');
+
+(async () => {
+  const browser = await chromium.launch(); // Open the browser
+  const page = await browser.newPage();
+
+  await page.goto('https://example.com'); // Navigate to desired page
+  await page.click('button#submit'); // Perform a click
+
+  const pageTitle = await page.title(); // Extract data from Page
+
+  console.log('Page Title:', pageTitle);
+
+  await browser.close();
+})();
+```
+
+## Headless Mode
+Headless mode means that you are able to run a web browser without a graphical user interface (GUI). Specifically, the browser operates in the background, allowing you to test, and gather data without displaying the browser window.
```

Comment:
The description of the headless mode could be clarified for readers who may not be familiar with the concept. Consider adding a more in-depth explanation of what headless mode means and give examples on why it is useful for automated testing

Reply:
Yep, will do